### DISCUSSION TITLE
Strange Warnings using the newer versions of tflint

### DISCUSSION THREAD

**Discussion by nvima** [terraform-linters/tflint#1548] [Q&A] [answered]
Hello, 
Since I updated my tflint, I get more warnings, which is actually very good. For example, I gave my variables a type.
However, in many modules I get the message "terraform missing version constraint for provider". Although I specified the version in the bottom main.tf, I never had that with the older tflint version.
Do I need to add version information in each module, or have I misconfigured tflint?
The Screenshot shows a split screen. The above main.tf is the /main.tf root file and the bottom main.tf is the /modules/cloudfront/main.tf 

![screenshot](https://user-images.githubusercontent.com/91621087/195982478-96e13de1-20ef-45b2-abd9-9b775ec0b169.png)

**Comment by bendrucker** [comment, marked as answer]
The rule in question tries to implement this official guidance:

https://www.terraform.io/language/expressions/version-constraints#terraform-core-and-provider-versions

In other words, yes every module should have version constraints. Child modules can have looser constraints (minimum version only) while parents will often specify at least a minor version, if not an exact version. 

While often you could leave out requirements from a child module with no effect as long as the parent sets them, this would mean that the module is no longer reusable or a valid root module. In order to run `terraform init` among other commands you need a valid root module so generally it's best to give that treatment to every module, especially if you want to run TFLint on those modules directly.
